plutus-ledger-api/testlib/PlutusLedgerApi/Test/EvaluationContext.hs

```diff
@@ -1,29 +1,119 @@
+
+-- | Example values of costs for plutus-v2, in expected ledger order, i.e. can be used in testing
+-- `V2.mkEvaluationContext . fmap snd`
+v2_costModelParamsForTesting :: [(V2.ParamName,Integer)]
+v2_costModelParamsForTesting = Map.toList $ fromJust $ extractCostModelParamsLedgerOrder v2_liftedCostModel
+
+-- | Example values of costs for plutus-v1, *in expected ledger order*, i.e. can be used in testing
+-- `V1.mkEvaluationContext . fmap snd`
+v1_costModelParamsForTesting :: [(V1.ParamName,Integer)]
+v1_costModelParamsForTesting = Map.toList $ fromJust $ extractCostModelParamsLedgerOrder v1_liftedCostModel
+
+v3_liftedCostModel, v2_liftedCostModel, v1_liftedCostModel
+    :: CostModel (CekMachineCostsBase Maybe)  (BuiltinCostModelBase MCostingFun)
+v3_liftedCostModel = liftToJust defaultCekCostModel
+v2_liftedCostModel = v3_liftedCostModel
+                    & machineCostModel
+                    %~ clearV3machineCostModel
+                    & builtinCostModel
+                    %~ clearV3builtinCostModel
+v1_liftedCostModel = v2_liftedCostModel
+                    & builtinCostModel
+                    %~ clearV2builtinCostModel
+
+-- Assign to `mempty` those builtins that the specific plutus version X introduces, and thus
+-- are missing from version X-1.
+-- Calling this function on a vX_BuiltinCostModel will turn it into a vX-1_BuiltinCostModel,
+-- by omitting the generation of costs concerning the missing builtins.
+clearV3builtinCostModel, clearV2builtinCostModel
+    :: (m ~ BuiltinCostModelBase MCostingFun) => m -> m
+clearV3builtinCostModel r = r
+               { paramBls12_381_G1_add = mempty
+               , paramBls12_381_G1_neg = mempty
+               , paramBls12_381_G1_scalarMul = mempty
+               , paramBls12_381_G1_equal = mempty
+               , paramBls12_381_G1_hashToGroup = mempty
+               , paramBls12_381_G1_compress = mempty
+               , paramBls12_381_G1_uncompress = mempty
+               , paramBls12_381_G2_add = mempty
+               , paramBls12_381_G2_neg = mempty
+               , paramBls12_381_G2_scalarMul = mempty
+               , paramBls12_381_G2_equal = mempty
+               , paramBls12_381_G2_hashToGroup = mempty
+               , paramBls12_381_G2_compress = mempty
+               , paramBls12_381_G2_uncompress = mempty
+               , paramBls12_381_millerLoop = mempty
+               , paramBls12_381_mulMlResult = mempty
+               , paramBls12_381_finalVerify = mempty
+               , paramKeccak_256 = mempty
+               , paramBlake2b_224 = mempty
+               }
+clearV2builtinCostModel r = r
+               { paramSerialiseData = mempty
+               , paramVerifyEcdsaSecp256k1Signature = mempty
+               , paramVerifySchnorrSecp256k1Signature = mempty
+               }
+
+-- Assign to `mempty` those CEK construct that the specific plutus version X introduces, and thus
```

Comment:
This is the ledger language `PlutusV<X>`, right?  `Constr` and `Case` appear in Plutus Core 1.0.1.  I guess that in fact the machine parameters depend on the Plutus Core language version and the buitlin parameters depend on the ledger language version, which is a pain (just like everything to do with these versions).

Reply:
You are semantically correct, but I see it more like "plutus version X introduces ledger language version Y which introduces those CEK machine constructs"